### DISCUSSION TITLE
Conditional select parameter on useQuery with type inference

### DISCUSSION THREAD

**Discussion by halileohalilei** [TanStack/query#8355] [Q&A] [answered]
I'm using v4 of TanStack Query for React on a project that integrates with a lot of different endpoints. For each endpoint, I create a file and implement a hook that uses `useQuery` to send a request. So for a `GET /orders` endpoint it looks something like this:

```ts
export function useOrders<T = GetOrdersResponse>(
  params: GetOrdersUrlParams,
  options: {
    isDisabled?: boolean;
    select?: (data: GetOrdersResponse) => T;
  } = {}
) {
  const { isDisabled = false, select } = options;

  return useQuery({
    queryKey: queryKeyFactory.getOrders(params),
    queryFn: () => fetchGetOrdersEndpoint(params),
    enabled: !isDisabled,
    select,
  });
}
```

This works great and all types are inferred properly when I pass a `select` with a different return type to `useOrders`. Now I want to provide some default abstraction to these hooks so that even if the consumer doesn't pass a `select` callback, the response is transformed into a default shape. So let's say I have the following function as a default select parameter:

```ts
function defaultSelect(data: GetOrdersResponse) {
  return data.rows;
}
```

When I  change the `select` parameter of `useQuery` to `select: select ?? defaultSelect`, I get this TS error on every parameter of `useQuery`:
```
No overload matches this call.
  The last overload gave the following error.
    Object literal may only specify known properties, and 'queryKey' does not exist in type 'readonly unknown[]'.ts(2769)
useQuery.d.ts(16, 25): The last overload is declared here.
```

I assume the return type of `select` can't be inferred correctly so it's inferring the whole options as `readonly unknown[]`. 

I know this is a bit of a niche use case, but any help or insight would be greatly appreciated.

**Comment by TkDodo** [comment]
I would do it like this:

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwFB1XArgZwKYEV3ZQCeANEgL5wBmUEIcA5AAIwCGAdpmwMYDWA9FGytuMALQBHAsQYBuJEirp2o4BHZwAJtiqt0AGxgBlbPuyiAFJtZsAXHADi2GAHko2qJgBK2TJE7YAJSIKHBCMOhQGtZsAHS0AO6Y8uQKMERg2I7Obh7evv44cAC8iGEQSfZOru6EmACqUPoACqxQrCCYcKnpmdk1eY0tbR1dpQhwYCMg9lxQwOwA5t0K2AAekLDUyqrqGDi5dQA8AColcD4RUScZ2Ee92BBUWjp6hiZmogB8XxahU+1OlUcrVPENWoDMGRUBAwDA1Jx7Mg0HBgJgACJo1gAIzMmgA-PZsRAIGYOPIUThPjBCXArDZWMCBnUfH51DhgsUvnAThTuucEKlgsi4Nx2fAJmjMZgcXjzrp9DgSHAqeZ4JRSrD4ez5KFwpENFg8NIiBYRagpIQiABpbBEewAbQY0DyDGVANGAF1oWhLcQAGLsewWTnc5q0EBo7DxXykgBu2DN5Uqk2mXXIgR9qGw7Fl2E09gAhFKsbj81mVaY1cGYqxQyEUSj9VFK9S4PjW2r6WxgvZtLoDMYq5Za4FQqhUhPAikFGLOPA-URzkbDp4kx6ZowqCSGN0x4vYrWkPx+I24AA9fGz8VwRcAJmXB1BmHX03sDG3EF35GVE1VohrBl61reIKkwWIN1iMwlhgAALPckHvQ8GWPU9G0vIA

**Reply by halileohalilei** [reply]
Thanks @TkDodo, I don't get any TS errors with this implementation, but `query2.data` is typed as `number | GetOrdersUrlParams | undefined` whereas I want it to be just `number | undefined`.

**Reply by halileohalilei** [reply, marked as answer]
I was able to achieve what I wanted with function overloads like [this](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwFB1XArgZwKYEV3ZQCeANCmjEWNnAKo76FEBK2m6ANjEgL5wBmUCCDgByAAIwAhgDtM0gMYBrAPRRsUhTAC0ARwLFRAbiRJ+6GVuAQZcACbZ+UzjADK2Dti0AKO1OkAXHAA4tgwAPJQDlCYrJiQctgAlIjk6jDoULZ+0gB0QgDumCY8ppTUIWGR0bFsCThwALyIcIWYQTLoIABGhADaALpwpeU0oRFRhJi0UBwAClJQUiCYTS1gi8tB8lDAMgDmw2VUNAAijs5crBlZACona9eZMvfUADyjEPz2Fy7unloAHwmJDYAAekFgAgsVhsGBw1Smb1ua3OThcTzuJ0B3nIGyWKyC40RMRm802KzIqAgYBg1jkAH4ggg4MBMKc2VJup47Ey4N0IBBPLIjHAcACYHzfP4pESqpMYnF6skmoC4LdRaUkkF6HgDCw2C5kcDQRDoPBzJY6XCsNgSZhcah8ct2pUJjUyQsCZgqXAada5MzWezOdzsLyggKhRoZKLxV5JUFpYE3falTYcClGmq0ZcYJiXidNWsEFrUqgFBn4Cy2RzMFyeWsnBwcCQxR4E2sHOiuP9O3xmv76cVTKh0s94XqmN5kGg4PomABpbBEIJ9UTQGqiNvOlYDX2oBfEABiMiTWbVcyEIDZ2HybCFADdsDPWhAimuAIxtgBMQx4SQHnA2AyA24ZBAAhLWoY8kB8ZaL6AElKYlZyPAR5EGstr2q+u5BKI-CCqIwxJEgGG5DkUhICoKhzqgAB6DIoVW876j+WEIgqDosnhYiERAxE8G2LLwTASaURecAAFKuOEAByuQ7Hs+zAPwRDJlI+TvpgKQAWRbEUTK1G0XRjFAA).
